File: Android/app/src/main/java/com/tsubuzaki/WoooshGo/discovery/DiscoveryController.kt
package com.tsubuzaki.WoooshGo.discovery

import android.content.Context
import android.net.nsd.NsdManager
import com.tsubuzaki.WoooshGo.peers.DeviceType
import com.tsubuzaki.WoooshGo.peers.PeerRegistry
import com.tsubuzaki.WoooshGo.settings.SettingsRepository
import java.security.SecureRandom
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch

/**
 * Wires settings -> advertiser and browser -> peer registry.
 *
 * [listenPort] is the core's bound QUIC port (`core.listenAddr()`), published verbatim in
 * the TXT `p` field (DESIGN.md §4); discovery does not start until it is non-zero.
 */
class DiscoveryController(
    context: Context,
    private val scope: CoroutineScope,
    private val settingsRepository: SettingsRepository,
    private val listenPort: () -> Int,
) {

    /** Rotating discovery ID: 8 random bytes, lowercase hex, new on every process start (PROTOCOL.md §3.1). */
    val rid: String = ByteArray(8)
        .also { SecureRandom().nextBytes(it) }
        .joinToString("") { "%02x".format(it) }

    private val appContext = context.applicationContext
    private val nsdManager = appContext.getSystemService(Context.NSD_SERVICE) as NsdManager

    val registry = PeerRegistry(scope)

    private val advertiser = DiscoveryAdvertiser(nsdManager)
    private val browser = DiscoveryBrowser(
        context = appContext,
        nsdManager = nsdManager,
        registry = registry,
        scope = scope,
        ownRid = { rid },
        ownServiceName = { advertiser.registeredServiceName },
    )

    /**
     * The TXT `dt` value. Platform-explicit per PROTOCOL.md §3.1 (the receiving UI cannot
     * tell a Pixel from an iPhone given only "phone"); sw600dp picks the form factor.
     */
    private val deviceType: DeviceType
        get() = if (appContext.resources.configuration.smallestScreenWidthDp >= 600) {
            DeviceType.ANDROID_TABLET
        } else {
            DeviceType.ANDROID_PHONE
        }

    private var started = false

    fun start() {
        if (started) return
        started = true

        scope.launch(Dispatchers.IO) { browser.start() }

        scope.launch(Dispatchers.IO) {
            settingsRepository.settings.collectLatest { settings ->
                // Display-name edits arrive per keystroke; don't flap the NSD
                // registration. apply()/stop() have no suspension points, so a newer
                // emission can only cancel this during the delay.
                delay(REREGISTER_DEBOUNCE_MS)
                advertiser.apply(
                    displayName = settings.displayName,
                    deviceType = deviceType,
                    visibility = settings.visibility,
                    rid = rid,
                    port = listenPort(),
                )
            }
        }
    }

    /** Explicit user refresh — the only in-process way the peer list is cleared. */
    fun refresh() {
        registry.clear()
        scope.launch(Dispatchers.IO) { browser.restart() }
    }

    private companion object {
        const val REREGISTER_DEBOUNCE_MS = 500L
    }
}

File: Windows/Wooosh/Discovery/DiscoveryController.cs
using System.Security.Cryptography;
using Microsoft.UI.Dispatching;
using Wooosh.Core;
using Wooosh.Peers;
using Wooosh.Settings;

namespace Wooosh.Discovery;

/// <summary><c>listenPort</c> is the core's bound QUIC port, published as the TXT <c>p</c>
/// field (DESIGN.md §4). Nothing is advertised until it is non-zero: announcing a port the
/// core is not listening on produces a device that answers nothing.</summary>
public sealed class DiscoveryController : IAsyncDisposable
{
    /// <summary>Display-name edits arrive per keystroke; re-registering on each one makes the
    /// device flicker in and out of every other device's list.</summary>
    private const int ReRegisterDebounceMs = 500;

    private readonly SettingsRepository _settings;
    private readonly Func<int> _listenPort;
    private readonly DnssdAdvertiser _advertiser = new();
    private readonly DnssdBrowser _browser;
    private readonly DispatcherQueueTimer _debounce;

    private bool _started;

    /// <summary>Rotating discovery ID (PROTOCOL.md §3.1), regenerated per process and
    /// deliberately not derived from the identity key: that is what stops a passive listener
    /// tracking this machine across networks.</summary>
    public string Rid { get; } = Convert.ToHexString(RandomNumberGenerator.GetBytes(8)).ToLowerInvariant();

    public PeerRegistry Registry { get; }

    public DiscoveryController(
        DispatcherQueue dispatcher,
        SettingsRepository settings,
        Func<int> listenPort)
    {
        _settings = settings;
        _listenPort = listenPort;
        Registry = new PeerRegistry(dispatcher);
        _browser = new DnssdBrowser(Registry, () => Rid, () => _advertiser.RegisteredInstanceName);

        _debounce = dispatcher.CreateTimer();
        _debounce.Interval = TimeSpan.FromMilliseconds(ReRegisterDebounceMs);
        _debounce.IsRepeating = false;
        _debounce.Tick += (_, _) => _ = ReadvertiseAsync();
    }

    public void Start()
    {
        if (_started)
        {
            return;
        }

        _started = true;
        _browser.Start();
        _settings.Changed += OnSettingsChanged;
        Readvertise();
    }

    public void Readvertise()
    {
        _debounce.Stop();
        _debounce.Start();
    }

    /// <summary>The only in-process way the append-only device list is cleared (DESIGN.md §5).</summary>
    public void Refresh()
    {
        Registry.Clear();
        _browser.Restart();
    }

    private void OnSettingsChanged() => Readvertise();

    private Task ReadvertiseAsync()
    {
        var settings = _settings.Current;
        var record = new TxtRecord
        {
            Rid = Rid,
            DisplayName = settings.DisplayName,
            DeviceType = DeviceType.Windows,
            Port = _listenPort(),
            Visibility = settings.Visibility,
        };

        return _advertiser.ApplyAsync(record);
    }

    public async ValueTask DisposeAsync()
    {
        _settings.Changed -= OnSettingsChanged;
        _debounce.Stop();
        _browser.Dispose();
        Registry.Dispose();
        await _advertiser.DisposeAsync();
    }
}
